### PR TITLE
add election class, process starts

### PR DESCRIPTION
**Pull request by rhsimplex**
Addresses #193.

### CODE REVIEW

**Comment by r-marques** [comment]
oh look a sunburst

**Comment by rhsimplex** [comment]
@r-marques that's how you know this is a quality PR

**Comment by r-marques** [comment]
hum not sure if that yellow and red should be there

File: bigchaindb/voter.py [hunk 1]
```diff
@@ -190,3 +190,69 @@ def start(self):
...
+            next_block = self.q_block_new_vote.get()
+            n_voters = len(next_block['block']['voters'])
+
+            vote_list = [vote['vote']['is_block_valid'] for vote in next_block['votes']]
+
+            n_invalid_votes = vote_list.count(False)
+
+            if n_invalid_votes > int(n_voters/2):
+                self.q_invalid_blocks.put(next_block)
+            else:
```
Comment (r-marques) [inline bigchaindb/voter.py]: I guess we can remove this.

**Comment by rhsimplex** [comment]
I don't know what it means. I ran the test suite locally and it seemed ok...

File: bigchaindb/voter.py [hunk 2]
```diff
@@ -190,3 +190,69 @@ def start(self):
...
+        """
+        Initialize the class with the needed queues.
+
+        Initialize a queue where blocks with new votes will be held
+        """
+        self.q_block_new_vote = q_block_new_vote
+        self.q_blocks_with_quorum = mp.Queue()
+        self.q_invalid_blocks = mp.Queue()
+
+    def check_for_quorum(self):
```
Comment (r-marques) [inline bigchaindb/voter.py]: I would move most of the logic in this method to `bigchaindb.core.Bigchain`.
It would be handy to have a method in which you give it a block (any block) and it returns either `valid`, `invalid`, or undecided.
We will need to use this functionality a lot.

File: bigchaindb/voter.py [hunk 3]
```diff
@@ -190,3 +190,69 @@ def start(self):
...
+            b = Bigchain()
+            # FIXME: this is unsafe and could lose transactions
+            #   Since queue items are removed immediately, there is no guarantee
+            #   all transactions will be rewritten. Imagine if node 5/5 casts a
+            #   deciding invalid vote.  This block will never show up in the
+            #   "update" changefeed again.  Suppose the below loop is writing
+            #   transactions and the worker crashes.  Those un-written transactions
+            #   will never be reviewed again.
+            #
+            #   Ideally, queue item should be removed on completion, a la celery or
```
Comment (r-marques) [inline bigchaindb/voter.py]: This would be possible to do but we would need a persistent queue instead of an in memory queue.

On the other hand remember that this is a distributed system and all nodes will be doing this for the same block so that would only be a problem if the all nodes crashed

File: bigchaindb/voter.py [hunk 4]
```diff
@@ -190,3 +190,69 @@ def start(self):
...
+        """
+        while True:
+            next_block = self.q_block_new_vote.get()
+            n_voters = len(next_block['block']['voters'])
+
+            vote_list = [vote['vote']['is_block_valid'] for vote in next_block['votes']]
+
+            n_invalid_votes = vote_list.count(False)
+
+            if n_invalid_votes > int(n_voters/2):
```
Comment (r-marques) [inline bigchaindb/voter.py]: We need to be careful with a split vote situation. A block is valid if the majority `n/2 + 1` says its valid and we may have a even number of nodes in the federation.

Image the case in which the federation has 10 nodes, 5 vote valid and 5 vote invalid. We have a split vote situation.
To be on the safe side in such a situation we should consider the block invalid, reassign the transactions and let the system try again.

**Comment by r-marques** [comment]
The code looks really nice. Just some remarks and some edge cases that need to be addressed.
And then we need some tests. You can check the `block.py` or `voter.py` unit tests

File: bigchaindb/voter.py (hunk 2 above)
Comment (r-marques) [inline bigchaindb/voter.py]: Another important thing that needs to happen here is that it is not enough to count votes. We also need to validate the signature in the vote.

File: tests/db/test_voter.py [hunk 5]
```diff
@@ -324,9 +324,40 @@ def test_quorum(self, b):
...
+
+        # construct invalid block
+        test_block_2['block']['votes'] = [invalid_vote_2, invalid_vote_2, valid_vote_2, valid_vote_2]
+        q_block_new_vote.put(test_block_2)
+
+        election.start()
+        time.sleep(1)
+        election.kill()
 
+        # TEST GOES HERE??
```
Comment (rhsimplex) [inline tests/db/test_voter.py]: Not sure how to test if a particular transaction made it back into the backlog. If someone can show me, then I can finish this test.

File: bigchaindb/core.py [hunk 6]
```diff
@@ -485,3 +485,25 @@ def get_unvoted_blocks(self):
...
+        vote_list = [vote['vote']['is_block_valid']
+                     for vote in block['block']['votes']
+                     if self.consensus.verify_vote_signature(vote)]
+
+        # validate votes here
+
+        n_invalid_votes = vote_list.count(False)
+        n_valid_votes = vote_list.count(True)
+
+        if n_invalid_votes >= int(n_voters/2):
```
Comment (vrde) [inline bigchaindb/core.py]: In Python 3 you can use the `//` operator:
```python
3 // 2 == int(3 / 2)
```

File: bigchaindb/core.py [hunk 7]
```diff
@@ -485,3 +485,25 @@ def get_unvoted_blocks(self):
...
+            return 'invalid'
```
Comment (vrde) [inline bigchaindb/core.py]: Should we export `invalid`, `valid`, and `undecided` as constants? It'd be cleaner.

File: bigchaindb/core.py (hunk 6 above)
Comment (rhsimplex) [inline bigchaindb/core.py]: oh that is neat

Comment (vrde) [inline bigchaindb/core.py]: Also, `int` is equivalent to `math.floor` for positive numbers.

In this scenario:
 - `n_voters` is odd, let's say `5`
 - `vote_list` is `[True, True, False, False]`
 - `n_invalid_votes` is `2`
 - `int(n_voters/2)` is `2`

the function would return `invalid`, is that what we want?

An alternative approach is to add a new variable, called `quorum` (so it's more explicit), that is `math.ceil(n_votes / 2)` or `round(n_votes / 2)` and then branch against it.

File: bigchaindb/core.py (hunk 7 above)
Comment (rhsimplex) [inline bigchaindb/core.py]: I don't know! Is there an example somewhere else?

File: bigchaindb/core.py [hunk 8]
```diff
@@ -485,3 +485,25 @@ def get_unvoted_blocks(self):
...
+        vote_list = [vote['vote']['is_block_valid']
+                     for vote in block['block']['votes']
+                     if self.consensus.verify_vote_signature(vote)]
+
+        # validate votes here
+
+        n_invalid_votes = vote_list.count(False)
+        n_valid_votes = vote_list.count(True)
+
+        if n_invalid_votes >= n_voters // 2:
```
Comment (vrde) [inline bigchaindb/core.py]: Also, `int` is equivalent to `math.floor` for positive numbers.

In this scenario:
 - `n_voters` is odd, let's say `5`
 - `vote_list` is `[True, True, False, False]`
 - `n_invalid_votes` is `2`
 - `int(n_voters/2)` is `2`

the function would return `invalid`, is that what we want?

An alternative approach is to add a new variable, called `quorum` (so it's more explicit), that is `math.ceil(n_votes / 2)` or `round(n_votes / 2)` and then branch against it.

File: bigchaindb/util.py [hunk 9]
```diff
@@ -501,6 +501,26 @@ def get_hash_data(transaction):
...
+    Args:
+        signed_vote (dict): a vote with the `signature` included.
+
+    Returns:
+        bool: True if the signature is correct, False otherwise.
+    """
+
+    data = signed_vote.copy()
+
+    signature = data.pop('signature')
```
Comment (r-marques) [inline bigchaindb/util.py]: I guess there is not need for the `copy` or the `pop`.
`public_key.verify(serialize(signed_vote['vote']), signed_vote['signature'])` should be enough.

File: bigchaindb/util.py [hunk 10]
```diff
@@ -501,6 +501,26 @@ def get_hash_data(transaction):
     return crypto.hash_data(serialize(tx))
 
 
+def verify_vote_signature(signed_vote):
```
Comment (r-marques) [inline bigchaindb/util.py]: Verifying the signature is not enough. We also need to check if a signature was created by a valid voter, so we need to check if the `node_pubkey` is in the `voters`.

Maybe have another method called `validate` block that checks if the `node_pubkey` is valid and then check the signature.

File: bigchaindb/core.py (hunk 7 above)
Comment (vrde) [inline bigchaindb/core.py]: Usually constants are at a [module level](https://www.python.org/dev/peps/pep-0008/#constants) but in this case I think it's fine to keep them on a class level.

File: bigchaindb/util.py (hunk 10 above)
Comment (rhsimplex) [inline bigchaindb/util.py]: You mean `voters` as in the voter list that comes with the vote? Couldn't that be faked too?

File: bigchaindb/voter.py [hunk 11]
```diff
@@ -196,3 +196,69 @@ def start(self):
...
+class Election(object):
+
+    def __init__(self, q_block_new_vote):
+        """
+        Initialize the class with the needed queues.
+
+        Initialize a queue where blocks with new votes will be held
+        """
+        self.q_block_new_vote = q_block_new_vote
+        self.q_blocks_with_quorum = mp.Queue()
```
Comment (r-marques) [inline bigchaindb/voter.py]: This queue does not seem to be used.

File: tests/db/test_voter.py [hunk 12]
```diff
@@ -297,6 +297,120 @@ def test_updating_block_number_must_be_atomic(self):
...
+        test_block['block']['votes'] = [invalid_vote, invalid_vote]
+        assert b.block_election_status(test_block) == 'undecided'
+
+        test_block['block']['votes'] = [valid_vote, valid_vote, valid_vote]
+        assert b.block_election_status(test_block) == 'valid'
+
+        test_block['block']['votes'] = [invalid_vote, invalid_vote, invalid_vote]
+        assert b.block_election_status(test_block) == 'invalid'
+
+
```
Comment (r-marques) [inline tests/db/test_voter.py]: too many blank lines

File: tests/db/test_voter.py [hunk 13]
```diff
@@ -297,6 +297,120 @@ def test_updating_block_number_must_be_atomic(self):
...
+        #
+        test_block['block']['votes'] = [invalid_vote, invalid_vote]
+        assert b.block_election_status(test_block) == 'invalid'
+
+        # test unanimously valid block with one improperly signed vote -- should still succeed
+        test_block['block']['votes'] = [valid_vote, valid_vote, valid_vote, improperly_signed_valid_vote]
+        assert b.block_election_status(test_block) == 'valid'
+
+        # test unanimously valid block with two improperly signed votes -- should fail to have quorum
+        test_block['block']['votes'] = [valid_vote, valid_vote,
```
Comment (r-marques) [inline tests/db/test_voter.py]: This is a situation in which quorum will never be reached and we cannot have that situation.

To be on the safe side we should define that vote is invalid if:
- the vote is correct and it voted the block invalid
- the vote is not correct (invalid signature, ...)

Better to just requeue the transactions and try again then having the block forever stuck in an undecided state

File: tests/db/test_voter.py [hunk 14]
```diff
@@ -297,6 +297,120 @@ def test_updating_block_number_must_be_atomic(self):
...
+
+        # test split vote
+        test_block['block']['votes'] = [invalid_vote, invalid_vote, valid_vote, valid_vote]
+        assert b.block_election_status(test_block) == 'invalid'
+
+        # test undecided
+        test_block['block']['votes'] = [valid_vote, valid_vote]
+        assert b.block_election_status(test_block) == 'undecided'
+
+        # test partial quorum situations for odd numbers of voters
```
Comment (r-marques) [inline tests/db/test_voter.py]: since we are recreating a block here maybe we can put this rest into a new unittest just to keep them smaller and more readable. Also easier to pinpoint the failure if the test fails

File: bigchaindb/consensus.py [hunk 15]
```diff
@@ -89,6 +89,18 @@ def verify_signature(signed_transaction):
...
+        """Verify a cast vote.
+
+        Args:
+            signed_vote (dict): signed vote to verify
+
+        Returns:
+            bool: True if the votes's required signature data is present
+                and correct, False otherwise.
+        """
+        raise NotImplementedError
```
Comment (diminator) [inline bigchaindb/consensus.py]: I guess the abstractmethods don't require `NotImplementedError` when using `ABC`.

same goes for the other methods in this abstract class

File: bigchaindb/consensus.py [hunk 16]
```diff
@@ -89,6 +89,18 @@ def verify_signature(signed_transaction):
         """
         raise NotImplementedError
 
+    @abstractmethod
```
Comment (diminator) [inline bigchaindb/consensus.py]: maybe also include:

`@staticmethod` on top of `@abstractmethod`

File: bigchaindb/core.py [hunk 17]
```diff
@@ -485,3 +486,25 @@ def get_unvoted_blocks(self):
...
+        vote_list = [vote['vote']['is_block_valid']
+                     for vote in block['block']['votes']
+                     if self.consensus.verify_vote_signature(vote)]
+
+        # validate votes here
+
+        n_invalid_votes = vote_list.count(False)
+        n_valid_votes = vote_list.count(True)
+
+        if n_invalid_votes >= math.ceil(n_voters / 2):
```
Comment (diminator) [inline bigchaindb/core.py]: maybe a line of comment on why using the ceil/floor

File: bigchaindb/util.py [hunk 18]
```diff
@@ -501,6 +501,26 @@ def get_hash_data(transaction):
...
+        signed_vote (dict): a vote with the `signature` included.
+
+    Returns:
+        bool: True if the signature is correct, False otherwise.
+    """
+
+    data = signed_vote.copy()
+
+    signature = data.pop('signature')
+    public_key_base58 = signed_vote['node_pubkey']
```
Comment (diminator) [inline bigchaindb/util.py]: in the latest code we use `sk` and `vk`  (signing key/verifying key) as it stands closer to the crypto jargon

so `public_key_base58 => vk_base58`

File: tests/db/test_voter.py (hunk 13 above)
Comment (rhsimplex) [inline tests/db/test_voter.py]: **Any invalid vote make a block invalid**, am I understanding that correctly?

File: bigchaindb/consensus.py (hunk 15 above)
Comment (rhsimplex) [inline bigchaindb/consensus.py]: This is an issue for the refactor then?

File: bigchaindb/consensus.py (hunk 16 above)
Comment (rhsimplex) [inline bigchaindb/consensus.py]: Why?

Comment (diminator) [inline bigchaindb/consensus.py]: Pylint likes to have the signatures of inherited methods to be the same

File: bigchaindb/consensus.py (hunk 15 above)
Comment (diminator) [inline bigchaindb/consensus.py]: sure

File: bigchaindb/consensus.py (hunk 16 above)
Comment (rhsimplex) [inline bigchaindb/consensus.py]: I'm not sure I agree, but it should go in the refactor anyway.  It's probably out of the scope of this PR for me to change all the function signatures in `consensus.py`.

Comment (diminator) [inline bigchaindb/consensus.py]: put in refactor: of course!

any reason why you would disagree on this?

Not specifying the `@staticmethod` decorator might trick child classes that the first argument is a `self` but that's not the case

Comment (rhsimplex) [inline bigchaindb/consensus.py]: I don't disagree, I'm `undecided` =)

**Comment by r-marques** [comment]
:+1:

File: bigchaindb/core.py [hunk 19]
```diff
@@ -485,3 +487,30 @@ def get_unvoted_blocks(self):
             unvoted.pop(0)
 
         return unvoted
+
+    def block_election_status(self, block):
+        """
+        Tallies the votes on a block, and returns the status: valid, invalid, or undecided
```
Comment (vrde) [inline bigchaindb/core.py]: Minor but important for consistency, docstrings should be *imperative*.

File: bigchaindb/core.py [hunk 20]
```diff
@@ -485,3 +487,30 @@ def get_unvoted_blocks(self):
...
+        n_valid_votes = sum(vote_list)
+        n_invalid_votes = len(vote_list) - n_valid_votes
+
+        # The use of ceiling and floor is to account for the case of an
+        # even number of voters where half the voters have voted 'invalid'
+        # and half 'valid'. In this case, the block should be marked invalid
+        # to avoid a tie. In the case of an odd number of voters this is not
+        # relevant, since one side must be a majority.
+        if n_invalid_votes >= math.ceil(n_voters / 2):
+            return 'invalid'
```
Comment (vrde) [inline bigchaindb/core.py]: would be nice to put `invalid`, `valid`, and `undecided` under `Bigchain.BLOCK_INVALID = 'invalid'`, `Bigchaindb.BLOCK_VALID = 'valid'', so it's easier for a developer to use the return value of the method. Also, IDEs and editors with code completion will make use of it.

File: tests/db/test_voter.py [hunk 21]
```diff
@@ -297,6 +297,153 @@ def test_updating_block_number_must_be_atomic(self):
...
+                      for member in test_federation]
+
+        # fake "no" votes
+        invalid_vote = [member.vote(test_block, 'abc', False)
+                        for member in test_federation]
+
+        # fake "yes" votes with incorrect signatures
+        improperly_signed_valid_vote = [member.vote(test_block, 'abc', True) for
+                                        member in test_federation]
+        [vote['vote'].update(this_should_ruin_things='lol')
```
Comment (vrde) [inline tests/db/test_voter.py]: lol

Comment (ttmc) [inline tests/db/test_voter.py]: Maybe `lol` is an emoticon for a person with their hands up.

**Comment by r-marques** [comment]
@rhsimplex maybe update the tests to use the class constants in the asserts. Other then that I think its ready to merge :+1:
